openstudiocore/src/utilities/core/FilesystemHelpers.cpp

```diff
@@ -172,5 +187,80 @@ namespace openstudio {
     {
       return to_time_t(openstudio::filesystem::last_write_time(t_path));
     }
+
+    openstudio::path create_temporary_directory(const openstudio::path &basename)
+    {
+      // making this count static/atomic so that we reduce the chance of collisions
+      // on each run of the binary. This is threadsafe, with the atomic
+      static std::atomic<unsigned int> count = 0;
+      constexpr auto allowed_attempts = 1000;
+
+      const auto temp_dir = openstudio::filesystem::temp_directory_path();
+
+      int attempts{0};
+
+      while (attempts < allowed_attempts) {
+        // concat number to path basename, without adding a new path element
+        auto filename = basename;
+        filename += openstudio::toPath("-" + std::to_string(count++));
+        const auto full_pathname = temp_dir / filename;
+        // full_path_name = {temp_path}/{base_name}-{count++}
+
+        try {
+          if (openstudio::filesystem::create_directories(full_pathname)) {
+            // if the path was created, then we know it was created for us
+            return full_pathname;
+          }
+        } catch (...) {
+          // swallow exceptions, we don't care, we only care when it succeeds
+        }
+      }
+
+      // after too many attempts we never made a path
+      return {};
+    }
+
+
+    openstudio::path home_path()
+    {
+      const auto build_path = [](const auto & ... elem) {
+        auto path_elem = [](auto &missing, const auto & env_var) {
+          if (const auto var = std::getenv(env_var); var != nullptr) {
+            return toPath(var);
+          } else {
+            missing = true;
+            return openstudio::path{};
+          }
+        };
+
+        bool missing_element = false;
+
+        const auto path = (path_elem(missing_element, elem) / ...).lexically_normal();
+
+        if (!missing_element && openstudio::filesystem::is_directory(path)) {
+          return path;
+        } else {
+          return openstudio::path{};
+        }
+      };
+
+      if (const auto home = build_path("USERPROFILE"); !home.empty()) {
+        LOG_FREE(Info, "FilesystemHelpers", "home_path USERPROFILE: " << toString(home));
+        return home;
+      }
+
+      if (const auto home = build_path("HOMEDRIVE", "HOMEPATH"); !home.empty()) {
+        LOG_FREE(Info, "FilesystemHelpers", "home_path HOMEDRIVE/HOMEPATH: " << toString(home));
+        return home;
+      }
+
+      if (const auto home = build_path("HOME"); !home.empty()) {
+        LOG_FREE(Info, "FilesystemHelpers", "home_path HOME: " << toString(home));
+        return home;
+      }
+
+      LOG_FREE(Warn, "FilesystemHelpers", "home_path No Home Found");
+      return toPath("/");
```

Comment:
@lefticus is "~/" a better guess, at least on unix?

Reply:
I maintained the behavior with the existing `QDir::homeDir()`. We cannot use `~` because that expansion won't be done, as far as I understand it. This is probably why Qt doesn't use `~`

Reply:
Ok great, if you matched Qt's implementation that will be perfect.